testing/multi-env-test-engine/src/main/java/org/projectnessie/junit/engine/MultiEnvTestEngine.java

```diff
@@ -116,18 +118,53 @@ public TestDescriptor discover(EngineDiscoveryRequest discoveryRequest, UniqueId
                   for (TestDescriptor child : children) {
                     // Note: this also removes the reference to parent from the child
                     discoverResult.removeChild(child);
-                    envRoot.addChild(child);
+
+                    // Must check whether the test class is a multi-env test, because discovery
+                    // returns all test classes.
+                    ClassBasedTestDescriptor classBased = (ClassBasedTestDescriptor) child;
+                    boolean multi = registry().isMultiEnvClass(classBased);
+                    if (multi) {
+                      envRoot.addChild(child);
+                    }
+                  }
+                }
+              });
+
+      // Also execute all other tests via the MultiEnv test engine to get the "thread per
```

Comment:
That's actually probably better anyways, since the filtering done in `MultiEnvTestFilter` becomes straightforward and we know that _all_ tests are going to be executed by the multi-env engine. I like it 👍

Reply:
I'm not sure this is ideal for downstream projects. Not all tests may want to get the "thread per class" model by default.

Could we add an annotation to control this behaviour?

Reply:
I don't see any reason to add more complexity here.